kernel/cpu/src/aarch64.rs

```diff
@@ -0,0 +1,113 @@
+use cortex_a::registers::MPIDR_EL1;
+use tock_registers::interfaces::Readable;
+
+use core::fmt;
+
+/// A unique identifier for a CPU core.
+#[derive(Copy, Clone, Debug)]
+#[repr(transparent)]
+pub struct CpuId(u32);
+
+/// An equivalent to Option<CpuId>, which internally encodes None as
+/// `u32::MAX`, which is an invalid CpuId (bits [4:7] of affinity level
+/// 0 must always be cleared). This guarantees that it compiles down to
+/// lock-free native atomic instructions when using it inside of an atomic
+/// type like [`AtomicCell`], as u32 is atomic when running on ARMv8.
+#[derive(Clone, Copy)]
+#[repr(transparent)]
+pub struct OptionalCpuId(u32);
+
+/// A unique identifier for a CPU core.
+#[derive(Copy, Clone, Debug)]
+#[repr(transparent)]
+pub struct MpidrValue(u64);
+
+/// Returns the number of CPUs (SMP cores) that exist and
+/// are currently initialized on this system.
+pub fn cpu_count() -> u32 {
+    // The ARM port doesn't start secondary cores for the moment.
+    1
+}
+
+/// Returns the ID of the bootstrap CPU (if known), which
+/// is the first CPU to run after system power-on.
+pub fn bootstrap_cpu() -> CpuId {
+    // The ARM port doesn't start secondary cores for the moment,
+    // so the current CPU can only be the "bootstrap" CPU.
+    current_cpu()
+}
+
+/// Returns true if the currently executing CPU is the bootstrap
+/// CPU, i.e., the first procesor to run after system power-on.
+pub fn is_bootstrap_cpu() -> bool {
+    // The ARM port doesn't start secondary cores for the moment,
+    // so the current CPU can only be the "bootstrap" CPU.
+    true
+}
+
+/// Returns the ID of the currently executing CPU.
+pub fn current_cpu() -> CpuId {
+    MpidrValue(MPIDR_EL1.get() as u64).into()
+}
+
+impl CpuId {
+    /// Reads an affinity level from this CpuId
+    ///
+    /// Valid affinity levels are 0, 1, 2, 3
+    pub fn affinity(self, level: u8) -> u8 {
+        assert!(level < 4, "Valid affinity levels are 0, 1, 2, 3");
+
+        (self.0 >> (level * 8)) as u8
+    }
+}
+
+impl MpidrValue {
+    pub unsafe fn get(self) -> u64 {
```

Comment:
I wanted to indicate that should not be used often; I forgot to document that by the way

Reply:
New doc:
```rust
    /// Obtain the inner raw u64 that was read from the MPIDR_EL1 register
    ///
    /// This is not dangerous, the unsafe modifier is simply there to
    /// indicate that this value should only be used when interacting with
    /// hardware/peripherals, in driver code for instance; it should not be
    /// used to identify/reference a CPU core.
```

Reply:
ok i see, then that's not the proper usage of unsafe. Getting an inner raw value is canonically safe in all Rust libraries, e.g., obtaining a pointer to something. _Using_ that raw value to re-obtain a higher-typed value is what's unsafe, e.g., dereferencing a pointer into a reference, or in this case, converting a raw integer into a `MpidrValue`.

We don't and shouldn't offer a public-facing way to convert a raw integer into a `MpidrValue`, so no unsafety is needed.

Rather than using unsafe, the way to indicate that a raw value should not be used to identify a CPU core is to _prevent_ it from happening by accepting only a higher-typed value that originated from a getter function, not a raw value.

Reply:
> Rather than using unsafe, the way to indicate that a raw value should not be used to identify a CPU core is to _prevent_ it from happening by accepting only a higher-typed value that originated from a getter function, not a raw value.

I agree, however driver crates will need to access that value...

I will remove the unsafe modifier.

Reply:
> I agree, however driver crates will need to access that value...

Right, and they can do some completely safely when obtaining that value from a bonafide instance of `MpidrValue` or something that was constructed from it, e.g., `CpuId`. This is the underlying principle of how type-carried invariants work, we're not really reinventing the wheel here.

> I will remove the unsafe modifier.

Great, thanks. Resolved.